src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs

```diff
@@ -326,14 +342,194 @@ public async Task<GetTransactionModel> GetTransactionAsync(string txid)
             return model;
         }
 
+        [ActionName("listaddressgroupings")]
+        [ActionDescription("Returns a list of grouped addresses which have had their common ownership made public by common use as inputs or as the resulting change in past transactions.")]
+        public AddressGroupingModel[] ListAddressGroupings()
+        {
+            if (!this.nodeSettings.TxIndex)
+                throw new RPCServerException(RPCErrorCode.RPC_INVALID_REQUEST, $"{nameof(ListAddressGroupings)} is incompatible with transaction indexing turned off (i.e. -txIndex=0).");
+
+            //Get all the address groupings.
+            var addressGroupings = GetAddressGroupings();
+
+            var addressGroupingModels = new List<AddressGroupingModel>();
+
+            foreach (var addressGrouping in addressGroupings)
+            {
+                foreach (var address in addressGrouping)
+                {
+                    var balance = this.walletManager.GetAddressBalance(address);
+                    var addressGroupingModel = new AddressGroupingModel()
+                    {
+                        Address = address,
+                        Amount = balance.AmountConfirmed
+                    };
+
+                    addressGroupingModels.Add(addressGroupingModel);
+                }
+            }
+
+            return addressGroupingModels.ToArray();
+        }
+
+        /// <summary>
+        /// Please see https://github.com/bitcoin/bitcoin/blob/726d0668ff780acb59ab0200359488ce700f6ae6/src/wallet/wallet.cpp#L3641
+        /// </summary>
+        /// <returns>A list of base 58 addresses.</returns>
+        private List<List<string>> GetAddressGroupings()
```

Comment:
Maybe this should go in WalletManager()? This is some cool new functionality that we might want to expose via api (in addition to RPC) so if the meat of the logic goes in wallet manager then it's easy to do. Just a suggestion! This is all very cool.

Reply:
Good call @mikedennis! I will move the logic 👍